packages/compiler-cli/src/ngtsc/annotations/src/ng_module.ts

```
@@ -177,39 +177,39 @@ export class NgModuleDecoratorHandler implements DecoratorHandler<NgModuleAnalys
-    // Unwrap ModuleWithProviders for modules that are locally declared (and thus static resolution
-    // was able to descend into the function and return an object literal, a Map).
-    if (entry instanceof Map && entry.has('ngModule')) {
-      entry = entry.get('ngModule') !;
+  /**
+   * Compute a list of `Reference`s from a resolved metadata value.
+   */
+  private resolveTypeList(resolvedList: ResolvedValue, name: string): Reference[] {
+    const refList: Reference[] = [];
+    if (!Array.isArray(resolvedList)) {
+      throw new Error(`Expected array when reading property ${name}`);
     }
 
-    if (Array.isArray(entry)) {
-      // Recurse into nested arrays.
-      refList.push(...resolveTypeList(entry, name));
-    } else if (entry instanceof Reference) {
-      if (!entry.expressable) {
-        throw new Error(`Value at position ${idx} in ${name} array is not expressable`);
-      } else if (!ts.isClassDeclaration(entry.node)) {
-        throw new Error(`Value at position ${idx} in ${name} array is not a class declaration`);
+    resolvedList.forEach((entry, idx) => {
+      // Unwrap ModuleWithProviders for modules that are locally declared (and thus static
+      // resolution was able to descend into the function and return an object literal, a Map).
+      if (entry instanceof Map && entry.has('ngModule')) {
+        entry = entry.get('ngModule') !;
       }
-      refList.push(entry);
-    } else {
-      // TODO(alxhub): expand ModuleWithProviders.
-      throw new Error(`Value at position ${idx} in ${name} array is not a reference: ${entry}`);
-    }
-  });
 
-  return refList;
+      if (Array.isArray(entry)) {
+        // Recurse into nested arrays.
+        refList.push(...this.resolveTypeList(entry, name));
+      } else if (entry instanceof Reference) {
+        if (!entry.expressable) {
+          throw new Error(`Value at position ${idx} in ${name} array is not expressable`);
+        } else if (!this.reflector.isClass(entry.node)) {
```

if we add an additional `ts.isDeclaration(entry.node)` here  then we don't need to change the signature of `isClass`.

---

But this will potentially hide errors (where `entry.node` is a Declaration but not a class) and not help in other places where `isClass()` might be called with the inner function declaration on ESM5.

Generally, I think this should be the reflector's concern.

---

I looked again and I agree that we need to change the `isClass` to take a `ts.Node`.